### PR TITLE
feat: allow compaction and persistence to retun no chunk (#2664)

### LINKED ISSUE #2546: Not throw error when all of rows of a OS or RUB chunk are soft deleted
# Rationale
In the PR #2522 I have discovered an [error will be thrown here](https://github.com/influxdata/influxdb_iox/blob/74d3c2e6d23cae7c2bcde7cc497ea77eb6062139/server/src/db/lifecycle/persist.rs#L147) if all rows of an OS chunk are soft deleted. We will need to change this behavior.

Similarly, this to [RUB here](https://github.com/influxdata/influxdb_iox/blob/cbf166c035c5c28f7d359896ed07e806d3a2a48e/server/src/db/lifecycle/move_chunk.rs#L68)

### LINKED ISSUE #2664: Compaction Return No Chunk
Currently lifecycle actions such as compact_chunks must return a DbChunk, specifically `Result<Arc<DbChunk>>>`. In order to resolve #2546 we wish for compaction to be able to return no chunk in the scenario when all rows have been deleted. The return type should therefore be changed to `Result<Option<Arc<DbChunk>>>>`

### PR DESCRIPTION
**Pull request by tustvold**
Mostly plumbing, this PR allows the persistence and compaction lifecycle actions to return no chunk in the event that all rows are deleted.

The way the test for the persist lifecycle action has to be written makes me very sad, I started fixing the plumbing to write this in a slightly less hacky way, but it's taking a bit longer than I had anticipated and I understand there is some urgency to getting this in, so hacky it is :sweat_smile: 

Closes #2664 
Closes #2546

### CODE REVIEW

File: server/src/db/lifecycle/persist.rs [hunk 1]
```diff
@@ -265,4 +276,88 @@ mod tests {
...
+            .unwrap();
+
+        assert!(maybe_chunk.is_none());
+
+        let chunks: Vec<_> = partition.read().chunk_summaries().collect();
+        assert_eq!(chunks.len(), 1);
+        assert_eq!(chunks[0].storage, ChunkStorage::ReadBuffer);
+        assert_eq!(chunks[0].row_count, 1);
+
+        // Add a third set of writes one of which overlaps the above chunk
```
Comment (tustvold) [inline server/src/db/lifecycle/persist.rs:322]: If it doesn't overlap the persist action just ignores it :laughing:

File: query_tests/src/scenarios.rs [hunk 2]
```diff
@@ -1101,15 +1108,13 @@ impl DbSetup for ChunkOrder {
...
         // transform chunk 0 into chunk 2 by persisting
-        let new_chunk = db
-            .persist_partition(
-                "cpu",
-                partition_key,
-                Instant::now() + Duration::from_secs(1),
-            )
-            .await
-            .unwrap();
-        assert!(lockable_chunk.id < new_chunk.id());
```
Comment (alamb) [inline query_tests/src/scenarios.rs:1112]: is there a reason to stop comparing the chunk ids? I am just wondering as it seems unrelated to this PR  (most other changes add a new `unwrap()`

File: server/src/db/lifecycle/compact.rs [hunk 3]
```diff
@@ -103,45 +105,46 @@ pub(crate) fn compact_chunks(
...
-        let throughput = (input_rows as u128 * 1_000_000_000) / elapsed.as_nanos();
-
-        info!(input_chunks=query_chunks.len(), rub_row_groups=rb_row_groups,
-                input_rows=input_rows, output_rows=guard.table_summary().total_count(),
-                sort_key=%key_str, compaction_took = ?elapsed, fut_execution_duration= ?fut_now.elapsed(),
-                rows_per_sec=?throughput,  "chunk(s) compacted");
-
-        Ok(DbChunk::snapshot(&guard))
+            None => {
+                info!(%addr, ?chunk_ids, "no rows to persist");
```
Comment (alamb) [inline server/src/db/lifecycle/compact.rs:144]:
```suggestion
                info!(%addr, ?chunk_ids, "no rows to persist, no chunk created");
```

File: server/src/db/lifecycle/compact.rs [hunk 4]
```diff
@@ -103,45 +105,46 @@ pub(crate) fn compact_chunks(
...
-        // input rows per second
-        let throughput = (input_rows as u128 * 1_000_000_000) / elapsed.as_nanos();
-
-        info!(input_chunks=query_chunks.len(), rub_row_groups=rb_row_groups,
-                input_rows=input_rows, output_rows=guard.table_summary().total_count(),
-                sort_key=%key_str, compaction_took = ?elapsed, fut_execution_duration= ?fut_now.elapsed(),
-                rows_per_sec=?throughput,  "chunk(s) compacted");
-
-        Ok(DbChunk::snapshot(&guard))
+            None => {
```
Comment (alamb) [inline server/src/db/lifecycle/compact.rs:143]: It doesn't really matter, but one thought i had while reading this code is that you might be able to an extra level of nesting (and make the diff smaller) by using a construct like

```rust
let rb_chunk = match maybe_rb_chunk {
  Some(rb_chunk) => rb_chunk,
  None => {
    info!(%addr, ?chunk_ids, "no rows to persist, no chunk created");
    return
  }
};
```

File: server/src/db/lifecycle/compact.rs [hunk 5]
```diff
@@ -217,4 +221,31 @@ mod tests {
...
+        let db = make_db().await.db;
+
+        write_lp(db.as_ref(), "cpu,tag1=cupcakes bar=1 10").await;
+        write_lp(db.as_ref(), "cpu,tag1=cupcakes bar=3 23").await;
+        write_lp(db.as_ref(), "cpu,tag1=cupcakes bar=2 26").await;
+
+        let partition_keys = db.partition_keys().unwrap();
+        assert_eq!(partition_keys.len(), 1);
+
+        // Cannot simply use empty predicate (#2687)
```
Comment (alamb) [inline server/src/db/lifecycle/compact.rs:236]: and now I see where #2687  came from lol

File: server/src/db/lifecycle/compact.rs [hunk 6]
```diff
@@ -217,4 +221,31 @@ mod tests {
...
+            .build();
+
+        // Delete everything
+        db.delete("cpu", Arc::new(predicate)).await.unwrap();
+        let chunk = db
+            .compact_partition("cpu", partition_keys[0].as_str())
+            .await
+            .unwrap();
+
+        assert!(chunk.is_none());
```
Comment (alamb) [inline server/src/db/lifecycle/compact.rs:249]: 👍

File: server/src/db/lifecycle/persist.rs [hunk 7]
```diff
@@ -132,27 +132,33 @@ where
...
                 );
             }
 
-            // NGA todo: we hit this error if there are rows but they are deleted
-            // Need to think a way to handle this (https://github.com/influxdata/influxdb_iox/issues/2546)
-            let to_persist = to_persist.expect("should be rows to persist");
+            let to_persist = match to_persist {
+                Some(to_persist) => to_persist,
+                None => {
+                    info!(%addr, ?chunk_ids, "no rows to persist");
```
Comment (alamb) [inline server/src/db/lifecycle/persist.rs:158]:
```suggestion
                    info!(%addr, ?chunk_ids, "no rows to persist, no chunks created");
```

File: server/src/db/lifecycle/persist.rs [hunk 8]
```diff
@@ -132,27 +132,33 @@ where
...
+                    return Ok(None);
+                }
+            };
```
Comment (alamb) [inline server/src/db/lifecycle/persist.rs:161]: There are several other things that `write_chunk_to_object_store` does such as updating the persistence windows and marking the chunk as "written to object store" (aka `guard.set_written_to_object_store(parquet_chunk)`)"

I wonder what are the implications of skipping these steps if there were no rows? 

The tests below show that somehow the source chunks get dropped, but I do still wonder about the persistence windows

File: server/src/db/lifecycle/persist.rs [hunk 9]
```diff
@@ -265,4 +276,88 @@ mod tests {
...
+        // Try to persist third set of writes
+        let maybe_chunk = db
+            .persist_partition("cpu", partition_key.as_str(), t2 + late_arrival)
+            .await
+            .unwrap();
+
+        assert!(maybe_chunk.is_none());
+
+        // The already persisted chunk should remain
+        let chunks: Vec<_> = partition.read().chunk_summaries().collect();
```
Comment (alamb) [inline server/src/db/lifecycle/persist.rs:358]: I see that the source ReadBuffer chunk is removed as expected, which is good. I am not 100% how this happens, but it seems to be the correct behavior

**Review by alamb — APPROVED** [review]
I am not sure if the way the persist logic is short-circuited leaves the persist windows in a bad state. The tests in this PR demonstrate that a chunk that is persisted but the persistence has no rows is dropped, which is the expected behavior but I did not expect given the code (meaning I don't fully understand this). 


Otherwise this PR looks great to me 👍

File: server/src/db/lifecycle/persist.rs (hunk 8 above)
Comment (tustvold) [reply server/src/db/lifecycle/persist.rs:161]: This is a good point, when the flush handle is dropped it isn't actually flushing out the persistable window. The tests pass because there is still an additional row to persist so what gets "orphaned" does eventually get flushed, but this may not always be the case.

Should be a simple fix to commit the flush handle, but very good spot! 👍

File: server/src/db/lifecycle/compact.rs (hunk 6 above)
Comment (NGA-TRAN) [reply server/src/db/lifecycle/compact.rs:249]: Nice

File: server/src/db/lifecycle/persist.rs (hunk 7 above)
Comment (NGA-TRAN) [reply server/src/db/lifecycle/persist.rs:158]: Very nice. Thanks @tustvold

**Review by NGA-TRAN — APPROVED** [review]
This looks good to me except the delete persistent test. I think you may want to ask @crepererum to have a look on Monday to make sure we do not change anything unexpectedly.

File: server/src/db.rs [hunk 10]
```diff
@@ -3647,10 +3676,6 @@ mod tests {
         let partition_key = "part_d";
         write_lp(&db, "cpu,part=d row=40,selector=0i 40").await;
         write_lp(&db, "cpu,part=d row=41,selector=1i 41").await;
-        db.compact_partition(table_name, partition_key)
-            .await
-            .unwrap();
-
```
[the remark this answers is not in the document]
Comment (tustvold) [reply server/src/db.rs:3653]: The persist_partition on the line below will persist from all unpersisted chunks, i.e. it is an implicit "compaction". The compaction here makes no different to the resulting distribution of chunks, all the unpersisted data will end up in a single ReadBufferAndObjectStore chunk regardless of if there is a preceding compact_partition call

File: server/src/db.rs [hunk 11]
```diff
@@ -679,7 +679,7 @@ impl Db {
         partition_key: &str,
         now: Instant,
         f_parquet_creation_timestamp: F,
-    ) -> Result<Arc<DbChunk>>
+    ) -> Result<Option<Arc<DbChunk>>>
     where
         F: Fn() -> DateTime<Utc> + Send,
     {
```
Comment (crepererum) [inline server/src/db.rs:685]: GitHub doesn't allow me to comment outside the change-set, but [this](https://github.com/influxdata/influxdb_iox/pull/2700/files#diff-45906f4b20ff30f9970a08127270631ed8d53922585aefa8625a436827b39fddR729-R735) `ensure!` macro should probably be replaced with `return None;` (+ a test).

**Review by crepererum — APPROVED** [review]
Looks good except for the missing case in `persist_partition_with_timestamp`.

File: query_tests/src/scenarios.rs (hunk 2 above)
Comment (tustvold) [reply query_tests/src/scenarios.rs:1112]: I removed it as it was testing a constraint that isn't supposed to be part of the contract, the only guarantee is chunk IDs are unique, there order is arbitrary

File: server/src/db/lifecycle/compact.rs (hunk 4 above)
Comment (tustvold) [reply server/src/db/lifecycle/compact.rs:143]: Yeah I did that initially but put it in the wrong place, before the chunks were dropped, so I rejigged. But perhaps removing the return was overkill - will change

File: server/src/db.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tustvold) [reply server/src/db.rs:3653]: I'll back out this change and raise a separate PR to simplify these tests

File: query_tests/src/scenarios.rs (hunk 2 above)
Comment (alamb) [reply query_tests/src/scenarios.rs:1112]: I see -- makes sense.  I suppose with the introduction of `ChunkOrder` that the Id can/should be treated as arbitrary

File: server/src/db.rs (hunk 11 above)
Comment (tustvold) [reply server/src/db.rs:685]: Created #2712 as this is actually a broader issue
